### PR TITLE
ENH: Add replace ufunc for bytes and unicode dtypes

### PR DESCRIPTION
**Pull request by lysnikolaou**

### CODE REVIEW

**Comment by lysnikolaou** [comment]
@seberg @ngoldbaum Another somewhat hacky ufunc for `replace` here. What's your opinion on this?

File: numpy/_core/umath.py [hunk 1]
```diff
@@ -31,3 +33,42 @@
...
+    >>> a = np.array(["The dish is fresh", "This is it"])
+    >>> np.core.umath.replace(a, 'is', 'was')
+    array(['The dwash was fresh', 'Thwas was it'], dtype='<U19')
+    """
+    counts = count(x1, x2, 0, numpy.iinfo(numpy.int_).max)
+    buffersizes = str_len(x1) + counts * (str_len(x3)-str_len(x2))
+    max_buffersize = numpy.max(buffersizes)
+    out = numpy.empty(x1.shape, dtype=f"{x1.dtype.char}{max_buffersize}")
+    _replace_impl(x1, x2, x3, x4, out=out)
+    return out
```
Comment (seberg) [inline numpy/_core/umath.py:74]: Let's not move it here?

File: numpy/_core/umath.py [hunk 2]
```diff
@@ -31,3 +33,42 @@
...
+    >>> np.core.umath.replace(a, 'mango', 'banana')
+    array(['That is a banana', 'Monkeys eat bananas'], dtype='<U19')
+    >>> a = np.array(["The dish is fresh", "This is it"])
+    >>> np.core.umath.replace(a, 'is', 'was')
+    array(['The dwash was fresh', 'Thwas was it'], dtype='<U19')
+    """
+    counts = count(x1, x2, 0, numpy.iinfo(numpy.int_).max)
+    buffersizes = str_len(x1) + counts * (str_len(x3)-str_len(x2))
+    max_buffersize = numpy.max(buffersizes)
+    out = numpy.empty(x1.shape, dtype=f"{x1.dtype.char}{max_buffersize}")
```
Comment (seberg) [inline numpy/_core/umath.py:72]: The one caveat here is that for nathan's new string dtype (or user one) this doesn't make sense, and we should only do it for the NumPy builtin strings (maybe in principle also for some other fixed length ones, but that seems less important).

**Review by seberg — COMMENTED** [review]
I am generally OK with the hack here.  Yes, the casting safety is side-stepped, but since we don't expose `out` if it is wrapped, that isn't a huge issue.

File: numpy/_core/umath.py (hunk 1 above)
Comment (lysnikolaou) [reply numpy/_core/umath.py:78]: Sorry, but I'm not sure what this means.

Comment (lysnikolaou) [reply numpy/_core/umath.py:78]: Oh yeah, I'm removing it from there now.

**Comment by lysnikolaou** [comment]
Let's try to merge this before the 2.0 release candidate if possible.

**Comment by ngoldbaum** [comment]
Missed this, will try to look at it tomorrow.

At the last community meeting Chuck said something about the middle of January at the earliest, so we have some time.

Are you still planning to do an `np.strings` PR? If not I'm happy to try.

I'm also planning to try working on stringdtype ufuncs next week. I'll ping you if I run into trouble adapting your code to work with utf8.

**Comment by lysnikolaou** [comment]
> Are you still planning to do an np.strings PR? If not I'm happy to try.

Yup, I'll start working on it today.

**Comment by seiko2plus** [comment]
It seems this PR is responsible for the following test failure across all CI jobs:
```Bash
=================================== FAILURES ===================================
  _____________________ TestMethodsScalarValues.test_replace _____________________
  
  self = <numpy._core.tests.test_defchararray.TestMethodsScalarValues object at 0x7f27b5cde3a0>
  
      def test_replace(self):
  >       assert_equal(np.char.replace('Python is good', 'good', 'great'),
                       'Python is great')
  
  self       = <numpy._core.tests.test_defchararray.TestMethodsScalarValues object at 0x7f27b5cde3a0>
  
  numpy/_core/tests/test_defchararray.py:736: 
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
  
  a = 'Python is good', old = 'good', new = 'great', count = 9223372036854775807
  
      @array_function_dispatch(_replace_dispatcher)
      def replace(a, old, new, count=None):
          """
          For each element in `a`, return a copy of the string with all
          occurrences of substring `old` replaced by `new`.
      
          Calls :meth:`str.replace` element-wise.
      
          Parameters
          ----------
          a : array-like of str or unicode
      
          old, new : str or unicode
      
          count : int, optional
              If the optional argument `count` is given, only the first
              `count` occurrences are replaced.
      
          Returns
          -------
          out : ndarray
              Output array of str or unicode, depending on input type
      
          See Also
          --------
          str.replace
      
          Examples
          --------
          >>> a = np.array(["That is a mango", "Monkeys eat mangos"])
          >>> np.char.replace(a, 'mango', 'banana')
          array(['That is a banana', 'Monkeys eat bananas'], dtype='<U19')
      
          >>> a = np.array(["The dish is fresh", "This is it"])
          >>> np.char.replace(a, 'is', 'was')
          array(['The dwash was fresh', 'Thwas was it'], dtype='<U19')
      
          """
          max_int64 = numpy.iinfo(numpy.int64).max
          count = count if count is not None else max_int64
      
          counts = numpy._core.umath.count(a, old, 0, max_int64)
          buffersizes = (
              numpy._core.umath.str_len(a)
              + counts * (numpy._core.umath.str_len(new) -
                          numpy._core.umath.str_len(old))
          )
          max_buffersize = numpy.max(buffersizes)
  >       out = numpy.empty(a.shape, dtype=f"{a.dtype.char}{max_buffersize}")
  E       AttributeError: 'str' object has no attribute 'shape'
```

**Comment by ngoldbaum** [comment]
Thanks! Should be fixed by #25484

**Pull request merged**